mephisto/data_model/agent.py

```diff
@@ -9,6 +9,7 @@
 from queue import Queue
 from mephisto.tools.misc import warn_once
 from uuid import uuid4
+from prometheus_client import Gauge
```

Comment:
is there any way to use a plugin architecture-type pattern here so that the prometheus dependency is opt-in and not a hard requirement. Perhaps we can have Gauge also be a noop if prometheus isn't found? I ask this to push these core data_model classes to be as lean and dependency-free as possible. I do understand the need for adding the prometheus_client here though, but perhaps it shouldn't be a rigid requirement for running Mephisto?

Reply:
I'm of two minds on this: On the one hand, I agree that we shouldn't be introducing dependencies throughout that aren't required for Mephisto to run. On the other hand, I think the value add of this kind of monitoring (especially once we add alerts and start having health checks and dependencies on these metrics (for instance, automatic scaling of `max_num_concurrent_units` based on server health)) means it _should_ be the default, and I worry making it opt-in would mean that the shortest path to average-case functionality is harder.

Perhaps there's a way to make it _opt-out_? I'm not sure what the build pattern for something like this is... But certainly I could move the prometheus classes into `mephisto.utils.metrics` and create no-op wrappers there if imports fail (though _that_ may also be complicated, as `Gauge` and other classes are classes and functions and also `with` contexts as well as decorators...).

I do want to make sure I'm not making this decision just because making it optional is more complex however.

Reply:
I think I'm reluctantly OK with making Prometheus a first class dependency, as the dashboard does seem useful. I did some research a few days back. In some ways Prometheus works well here, but in others it's not a perrrrfect fit. In the ideal case, I think Prometheus likes to be running all the time while polling long-running tasks (usually dedicated service instances). For things that are short-lived "batch jobs" (which maybe is a task run? idk.) they mention using something called the [Pushgateway](https://prometheus.io/docs/practices/pushing/#when-to-use-the-pushgateway). Not sure if you've looked into this.

I was also trying to see if there's a way for a Mephisto run to output it's metrics to a data file, which could later be replayed into Prometheus, but I found nothing like this after a bunch of research. Am I correct here? Is there no way to do a "replay", i.e. persist/share logging data?

Reply:
> In some ways Prometheus works well here, but in others it's not a perrrrfect fit.

Agreed especially for the simple case of short pilot jobs, but logging every 5-10m over the course of a 3 day job, or (in the future) monitoring 3-4 different simultaneous jobs seems to work.

> For things that are short-lived "batch jobs" (which maybe is a task run? idk.) they mention using something called the [Pushgateway](https://prometheus.io/docs/practices/pushing/#when-to-use-the-pushgateway).

Yup did some reading on this - in the end I think going with a faster polling rate ends up better than the benefits that the `Pushgateway` really provides. From their docs: 'A "service-level" batch job is one which is not semantically related to a specific machine or job instance' - all of the metrics for a specific TaskRun have an associated 'job' on the Prometheus side. It also still requires a running server.

> Am I correct here? Is there no way to do a "replay", i.e. persist/share logging data?

There _is_, but it's a little silly (still intending to support it in a follow-up in 1.1 though): The prometheus *server* is able to do a dump that can be later imported, so the procedure for this would be to run a server during launch (which we automatically do) and then dump the metrics from the server that are tagged with the specific task run on close (which we don't do yet, but can).

Reply:
> Assuming Prometheus is a hard dependency for all Mephisto runs - how does this affect the user, if in any way? Is installing Prometheus first via install_mac.sh or install_linux.sh first a MUST? If so that is a severe detriment to the new user experience.

It's not a must - the system will run but skip logging. There will be a warning to this effect:
```
if not os.path.exists(PROMETHEUS_EXECUTABLE):
        logger.warning(
            f"Cannot collect and display metrics without installing metrics. "
            f"Use the scripts in the metrics dir ({METRICS_DIR}) to install. "
        )
        return False
```
I can probably soften this a little bit if you think it's necessary though.

> What is the story around when a user accesses the data and how? Can they access it after the task ends? Is there something they have to do before launching a task to opt-in to the metrics, or no? You touch a bit on this in your testing section, but i'm still unclear on what happens after the run ends? Is it too late at that point?

So I've set this up so that, if the user has run one of the metrics install scripts, upon launching a job we also launch the relevant server. This server will persist (and the data will persist after the job is shutdown), so then they can use the `view_metrics.py` script to see the results.

Reply:
Gotcha, so if I understand correctly: Once Prometheus is installed, the first run launches the Prometheus server. The server stays alive after the job ends, and then must be manually closed.

Correct me if I'm wrong: Subsequent jobs will reuse the existing running server if it's already up? If someone launches Job A and then Job B, we still differentiate the metrics instead of commingling right? What happens if people run multiple jobs at a time - do we support that easily or do they have to do some configuration?

Reply:
> If someone launches Job A and then Job B, we still differentiate the metrics instead of commingling right? 

Correct, upon completion of Job A, the metrics will be separate on Job B

> What happens if people run multiple jobs at a time - do we support that easily or do they have to do some configuration?

This will come in 1.1